Upgrade to Babel 6 and remove default from export

adrianleb opened the pull request:
Since the recent code changes when adding prefix styles, when running `npm prepublish` Babel was crashing on this line https://github.com/Khan/aphrodite/blob/master/src/util.js#L29 since it expects a plugin to be able to transpile said code.

As a consequence, the build published to NPM failed to contain the latest `/lib` files, if you right now install version `0.1.2` of aphrodite via NPM it will not be the latest according to what's on Github although the package.json version number indicates so (one good hint is the outdated README.md and complete lack of auto-prefixing code which is what made me investigate this problem)

With this PR I've added the necessary Babel plugin and a .babelrc file, so running `npm run prepublish` works as expected. the last step is to publish a new version to NPM so everyone has, for sure, the latest version of aphrodite including prefixed styles magic!

kevinbarabash commented:
The travis-ci build is failing probably because it's trying to use a babel 6 plugin with babel 5.  We should probably upgrade to babel 6.

adrianleb commented:
Right, I can upgrade to Babel 6 part of this PR so that travis-ci is happy.

Any reason to keep using Babel 5? I don't see any non-trivial change required.

kevinbarabash commented:
If you'd be willing to upgrade to Babel 6 that would be awesome.  If you haven't done so already, please sign our Contributor License Agreement: http://www.khanacademy.org/r/cla.

adrianleb commented:
CLA Signed ✅
Babel 6 upgrade ✅
Travis now passing ✅
One small change to be noted, on https://github.com/Khan/aphrodite/pull/27/files#diff-1fdf421c05c1140f6d71444ea2b27638R59 I've removed the `default` since the compiled result was not what was expected (and reflecting the docs) This article covers why http://www.2ality.com/2015/12/babel-commonjs.html#why_isnt_the_default_export_done_like_a_commonjs_single_export%3F

kevinbarabash commented:
Looks good.  I'm going let @xymostech or @jlfwong hit the merge button because they are the main contributors for this repo.

xymostech commented on .gitignore:
```diff
@@ -1,2 +1,3 @@
 lib
 node_modules
+.DS_Store
```
This should be in a global `.gitignore` and not be added here.

xymostech commented:
This all looks good to me @adrianleb! Good call on the `default` usage, I'm guessing this works for us because we just use the ES5 compiled version.

Could you squash your commits?

jlfwong commented:
Hi @adrianleb, thanks for the PR!

I'm a little confused about the scenario you originally ran into, however. The `npm prepublish` script should've used the version in `node_modules`, which should've been Babel 5 per the existing `package.json`. That script doesn't crash with Babel 5, right? So how was it using Babel 6 for you?

adrianleb commented:
Hey @jlfwong, no problem, my pleasure!

So the scenario is that when using `0.1.2` from NPM the code !== to what's on Github as 0.1.2, you can check the difference here:
https://raw.githubusercontent.com/Khan/aphrodite/master/dist/aphrodite.js
vs
https://npmcdn.com/aphrodite@0.1.2/dist/aphrodite.js

Notably, this means everything related to inline-prefix-styles.

After checking out master and trying to prepublish locally babel was getting stuck on the line I mentioned above (https://github.com/Khan/aphrodite/blob/master/src/util.js#L29)

I do have babel installed globally at version `6.5.1`

If prepublish works with the current version of babel and `{...restSpread}` out of the box on a machine different than mine then the fix to this issue is just another call to `npm publish` with the right code for `0.1.2`

jlfwong commented:
@adrianleb Thanks for the explanation -- the difference between what's on GitHub and what's on npm isn't surprising. It sounds like we should version bump and publish to put the style prefixing up on npm.

I'm still a little confused about the global babel installation -- I would've thought that npm scripts (e.g. if you had run `npm prepublish`) would set a PATH that forces the `node_modules/.bin` to take precedence over a global installation. I guess that might not be the case if you hadn't first run `npm install`, because the `node_modules/.bin` would be empty then?

If I understand correctly, this PR shouldn't be fixing any bugs, but just upgrading to Babel 6 (which is then, IMO, what this PR should be titled with). Please let me know if I'm still misundertanding

adrianleb commented:
@jlfwong You're absolutely correct!

I've tested again on a different computer and the version on master runs fine, I guess I had a bad conf on my machine.

In short, this PR becomes about upgrading to latest Babel & removing the `default` from the main export line which is not needed/not the best pattern.

Most important, a polite request to have the NPM version of aphrodite updated since that's what triggered me to do this. :)

I've updated the PR title and squashed the commits so If still fitting the code is ready to be merged.

jlfwong commented:
@adrianleb Thanks for investigating on another computer :)

I've published 0.2.0 now, which should contain the style prefixing changes. Thanks for letting us know it was out of date.

As for this diff, I'm not super inclined to jump on upgrading without cause at the moment -- other portions of Khan Academy's internal code uses babel 5 at the moment, so I'd rather leave this alone unless there are pressing needs.

When we do get around to upgrading though, this will serve as a great reference for other repositories, so I'm going to leave this open for now if that's alright with you.

adrianleb commented:
Thanks @jlfwong!

Agree updating Babel has way less urgency if everything works.

The one thing which this PR also cover which can be worth looking into is the `default` export.

Right now to use Aphrodite I'm having to do `import Aphrodite from 'aphrodite';` since according to recent es6 changes `import {Stylesheet, css} from 'aphrodite'` doesn't work if there's only a `default` export.

It's not exactly a bug but documentation can be a bit misleading about it.

jlfwong commented:
@adrianleb Oh, interesting! If Babel 5 will take the version without the `default`, then I'm happy to take that as a PR by itself.

lencioni commented on package.json:
```diff
@@ -26,9 +26,12 @@
-    "babel": "^5.8.23",
-    "babel-core": "^5.8.25",
-    "babel-loader": "^5.3.2",
+    "babel": "^6.5.2",
```
FWIW, I ended up doing a similar update in https://github.com/Khan/aphrodite/pull/281

kevinbarabash commented:
We've update to using babel 6 internally quite some time ago.

kevinbarabash commented:
It looks like Aphrodite was upgraded to use babel 6 as well some time ago so I'm going to close this.
